amundsen_application/static/js/components/Tags/TagInfo/index.tsx

```diff
@@ -1,12 +1,15 @@
 import * as React from 'react';
 import { Link } from 'react-router-dom';
 import { Tag } from '../types';
+import { log } from "../../../ducks/utilMethods";
+
 
 import './styles.scss';
 
 interface TagInfoProps {
   data: Tag;
   compact?: boolean;
+  location: string;
```

Comment:
I think we should look into isolation what is needed for logging away from the properties and behavior of the component as much as possible. Besides perhaps requiring an `id` and hooking up the component to call the `log` method when interacted with, can we try to isolate everything else into some kind of logging utils?

For example is any of this information available from the `event` or the `target`? Or perhaps what if we grabbed `window.location` in the log method?

Reply:
This currently isn't `window.location`, but something like "navbar",  "frequent users", or "column details". The page URL might also be a useful thing to log. Maybe we can add something in the generic handler to automatically add the URL.

Reply:
Even so -- beyond giving it the component the ability to call the log method I'm still wondering how we can avoid introducing new properties for things that are not inherent to the component's functional purpose.

Another idea...what if our ids were generated to have all of the extra information that don't relate the the component's functional purpose. Perhaps we could have utility methods to generate ids of a specific pattern, and parse out the information we need in the `log` method.

Reply:
I think the analytics events should be highly structured and predictable as someone or some system downstream will ultimately need to query these events. I agree that certain events will need some unique fields that don't really fit into any of the predefined fields, but I don't think the ID is the right place to put this data.

We could try something like the [fieldsObject] in the Google Analytics API listed below.

<img width="648" alt="Screen Shot 2019-04-09 at 11 01 07 AM" src="https://user-images.githubusercontent.com/4338702/55823743-41be0d00-5ab7-11e9-8a51-3da95d566dc7.png">


Personally I am in favor of killing our custom logger in favor of pushing all events into GA, but that is out of scope for now. As an alternative, we send events to both our custom logger as well as GA. I'd only need to modify the `logAction` function.

Reply:
Regarding predictability, the idea is that the information sent in the actual logging event would still be structured. `id`s would be generated according to a predefined structure. Information would also be parsed out of that `id` via some utility method, and the payload sent with the analytics event would still be the same -- the difference is the component itself would contain no actual logic for the logging, besides calling the log methods. For example why pass a string like "navbar" to the component if we can include it in the `id` and parse it out when we are creating the payload to send with the analytics event.

This is just one idea -- what I'm raising here is the concept of keeping our components clean and decoupled from what we need for logging, and brainstorming how to go about that. I'm wondering how we can avoid adding new properties that are unrelated to the component's purpose just so that we can pass it to the logger, because that approach will not scale well. I believe it could be possible to grab all of the information we need from passing just the DOM element and dealing with it accordingly outside of the component. 

It would be great if all the component had to do was leverage utilities like `logClick(element: HTMLElement)` and `logScroll(window:  Window?)`, and isolated logic that lives outside of the component would handle the rest.

Reply:
You're suggesting we take something like this
```
<a href={url}
   target='_blank'
   onClick={() => logAction({
     command: 'click',
     target_id: 'frequent-users',
     target_type: 'link',
     label: fullName,
     location: 'table-details',
   })}
>
```
and convert it to 

```
<a href={url}
   id=`frequent-users:link:${fullName}:table-details`
   target='_blank'
   onClick={logClick(this)}
>

logClick = (target: HTMLElement) => {
  let parts = target.id.split(':')
  logAction({
    command: "click",
    target_id: parts[0],
    target_type: parts[1],
    label: parts[2],
    location: parts[3],
  });
}
```

While this looks cleaner, it isn't not much different from defining a function like `logAction(action, target_id, target_type, label, location)`, with the added requirement that you need to memorize the ID structure. Another approach that many analytics libraries use (and the one I'm most familiar with) is to annotate the element with some analytics attributes.


```
<a
   href="#"
   data-metrics-event-name="commentPageClick"
   data-metrics-comment-id={commentId}
   data-metrics-page-num={currentPage + 1}>
   Next
</a>
```
from: https://github.com/nfl/react-metrics

```
<a href="file.pdf"
   analytics-on="click"
   analytics-event="Download"
   analytics-properties="{ category: 'Content Actions' }">Download</a>
```
from: https://github.com/angulartics/angulartics

Reply:
> While this looks cleaner, it isn't not much different from defining a function like `logAction(action, target_id, target_type, label, location)`, with the added requirement that you need to memorize the ID structure. Another approach that many analytics libraries use (and the one I'm most familiar with) is to annotate the element with some analytics attributes.

What I’m suggesting is more like:
```
<a href={url}
   id=generateIdWithParametersThatCantBeInferred()
   target='_blank'
   onClick={logClick(this)}
>

// lives in another file
logClick = (target: HTMLElement) => {
  logAction({
    command: "click" OR we can get this from event.type if passing the event and not just the target,
    target_id: either parsed from or the same as target.id,
    target_type: we can get this from the target.type,
    label: unless I misunderstand what this value is, we can also get this from somewhere in the target?
    location: while not the exact url, we can still parse ‘table_detail` from the url which if not on the target should be on the event,
  });
}
```
Generating and parsing the `id` would be handled by methods that take a payload of a specific interface. Developers would be guided by the TypeScript interface to know what to pass -- no memorization, only the need to know the interface just like everything else. (but this isn't the main point)

What I hope is bit clearer from above is that I think it may be possible to get many of those values by looking into the breadth of data that is automatically provided to us via the browser event APIs. If we don't go that exact route -- though it would be nice to have cleaner single line log methods instead --  the main concept I hope we can figure out is how to avoid adding one-off properties like `location`, or any other log properties in the future, to the component properties so that we can keep our components clean and only concerned with their function.